server/frontend.go

```diff
@@ -106,46 +123,70 @@ func (s *server) NVMeSubsystemStats(ctx context.Context, in *pb.NVMeSubsystemSta
-	log.Printf("Received from client: %v", in.GetControllerId())
+	log.Printf("Received from client: %v", in.ControllerId)
+	controller, ok := controllers[in.ControllerId.Value]
+	if !ok {
+		return nil, errors.New(fmt.Sprintf("error finding controller %s", in.ControllerId.Value))
+	}
+	delete (controllers, controller.Id.Value)
 	return &pb.NVMeControllerDeleteResponse{}, nil
 }
 
 func (s *server) NVMeControllerUpdate(ctx context.Context, in *pb.NVMeControllerUpdateRequest) (*pb.NVMeControllerUpdateResponse, error) {
-	log.Printf("Received from client: %v", in.GetController())
+	log.Printf("Received from client: %v", in.Controller)
+	controllers[in.Controller.Id.Value] = *in.Controller
 	return &pb.NVMeControllerUpdateResponse{}, nil
 }
 
 func (s *server) NVMeControllerList(ctx context.Context, in *pb.NVMeControllerListRequest) (*pb.NVMeControllerListResponse, error) {
-	log.Printf("Received from client: %v", in.GetSubsystemId())
-	Blobarray := make([]*pb.NVMeController, 3)
+	log.Printf("Received from client: %v", in.SubsystemId)
+	Blobarray := []*pb.NVMeController{}
+	for _, controller := range controllers {
+		Blobarray = append(Blobarray, &controller)
+	}
 	return &pb.NVMeControllerListResponse{Controller: Blobarray}, nil
 }
 
 func (s *server) NVMeControllerGet(ctx context.Context, in *pb.NVMeControllerGetRequest) (*pb.NVMeControllerGetResponse, error) {
-	log.Printf("Received from client: %v", in.GetControllerId())
-	return &pb.NVMeControllerGetResponse{Controller: &pb.NVMeController{Name: "Hello " + fmt.Sprint(in.GetControllerId())}}, nil
+	log.Printf("Received from client: %v", in.ControllerId)
+	controller, ok := controllers[in.ControllerId.Value]
+	if !ok {
+		return nil, errors.New(fmt.Sprintf("error finding controller %s", in.ControllerId.Value))
+	}
+	return &pb.NVMeControllerGetResponse{Controller: &controller}, nil
 }
 
 func (s *server) NVMeControllerStats(ctx context.Context, in *pb.NVMeControllerStatsRequest) (*pb.NVMeControllerStatsResponse, error) {
-	log.Printf("Received from client: %v", in.GetControllerId())
+	log.Printf("Received from client: %v", in.Id)
 	return &pb.NVMeControllerStatsResponse{}, nil
 }
 
 //////////////////////////////////////////////////////////
+var namespaces = map[string]pb.NVMeNamespace{}
 
 func (s *server) NVMeNamespaceCreate(ctx context.Context, in *pb.NVMeNamespaceCreateRequest) (*pb.NVMeNamespaceCreateResponse, error) {
 	log.Printf("NVMeNamespaceCreate: Received from client: %v", in)
+	subsys, ok := subsystems[in.Namespace.SubsystemId.Value]
+	if !ok {
+		err := errors.New(fmt.Sprintf("unable to find subsystem %s", in.Namespace.SubsystemId.Value))
+		log.Printf("error: %v", err)
+		return nil, err
+	}
+
 	params := NvmfSubsystemAddNsParams{
-		Nqn: in.GetNamespace().GetSubsystemId(),
+		Nqn: subsys.Nqn,
 	}
-	params.Namespace.BdevName = in.GetNamespace().GetBdev()
+	// TODO: using bdev as malloc until middlelayer reference is sorted out
```

Comment:
we have implantation today they relies on passing BDEV... 
so please leave the TODO comnent...
but let's use `CryptoKeyId.Value`  as a workaround to pass the desired BDEV... 
That's why I didn't want to delete the `bdev` from the front end...

Reply:
ack; fixed.